app/layer_generators/main.go

```diff
@@ -0,0 +1,222 @@
+							if idx := strings.Index(s, OPEN_TRACING_PARAMS_MARKER); idx != -1 {
+								for _, p := range strings.Split(s[idx+len(OPEN_TRACING_PARAMS_MARKER):], ",") {
+									paramsToTrace[strings.TrimSpace(p)] = true
+								}
+							}
+
+						}
+					}
+					if e.Params != nil {
+						for _, param := range e.Params.List {
+							for _, paramName := range param.Names {
+								paramType := (formatNode(src, param.Type))
+								params = append(params, Param{Name: paramName.Name, Type: paramType})
+							}
+						}
+					}
+					if e.Results != nil {
+						for _, result := range e.Results.List {
+							results = append(results, formatNode(src, result.Type))
+						}
+					}
+					metadata.Methods[methodName] = Method{Params: params, Results: results, ParamsToTrace: paramsToTrace}
+
+				}
+			}
+		}
+
+		return true
+	})
+	return metadata
+}
+
+func GenerateLayer(name, templateFile string) string {
+	out := bytes.NewBufferString("")
+	metadata := ExtractStoreMetadata()
+	metadata.Name = name
+
+	myFuncs := template.FuncMap{
+		"joinResults": func(results []string) string {
+			return strings.Join(results, ", ")
+		},
+		"joinResultsForSignature": func(results []string) string {
+			if len(results) == 0 {
+				return ""
+			}
+			if len(results) == 1 {
+				return strings.Join(results, ", ")
+			}
+			return fmt.Sprintf("(%s)", strings.Join(results, ", "))
+		},
+		"genResultsVars": func(results []string) string {
+			vars := []string{}
+			for idx := range results {
+				vars = append(vars, fmt.Sprintf("resultVar%d", idx))
+			}
+			return strings.Join(vars, ", ")
+		},
+		"errorToBoolean": func(results []string) string {
+			for idx, typeName := range results {
+				if typeName == "*model.AppError" {
+					return fmt.Sprintf("resultVar%d == nil", idx)
+				}
+			}
+			return "true"
+		},
+		"errorPresent": func(results []string) bool {
+			for _, typeName := range results {
+				if typeName == "*model.AppError" {
+					return true
+				}
+			}
+			return false
+		},
+		"errorVar": func(results []string) string {
+			for idx, typeName := range results {
+				if typeName == "*model.AppError" {
+					return fmt.Sprintf("resultVar%d", idx)
+				}
+			}
+			return ""
+		},
+		"joinParams": func(params []Param) string {
+			paramsNames := []string{}
+			for _, param := range params {
+				s := param.Name
+				if strings.HasPrefix(param.Type, "...") {
+					s += "..."
+				}
+				paramsNames = append(paramsNames, s)
+			}
+			return strings.Join(paramsNames, ", ")
+		},
+		"joinParamsWithType": func(params []Param) string {
+			paramsWithType := []string{}
+			for _, param := range params {
+				paramsWithType = append(paramsWithType, fmt.Sprintf("%s %s", param.Name, param.Type))
+			}
+			return strings.Join(paramsWithType, ", ")
+		},
+	}
+
+	t := template.Must(template.New("opentracing_layer.go.tmpl").Funcs(myFuncs).ParseFiles(templateFile))
+	err := t.Execute(out, metadata)
+	if err != nil {
+		panic(err)
+	}
+	return out.String()
```

Comment:
It seems like we are returning string and then again converting to []byte slice to the next function. What do you think about returning a byte slice instead, so that double allocations are avoided ?